### PR TITLE
RestartSerialization: add parallel support

### PR DESCRIPTION
**Pull request by akva2**
This extends the restart serialization to support parallel simulations.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: CMakeLists.txt [hunk 1]
```diff
@@ -91,6 +91,10 @@ endif()
 
 find_package(opm-common REQUIRED)
 
+if(USE_MPI)
+  set(HDF5_PREFER_PARALLEL TRUE)
+endif()
```
Comment (bska) [inline CMakeLists.txt:96]: This is probably the right thing to do in general, but CMake has an [issue](https://gitlab.kitware.com/cmake/cmake/-/issues/19117) that's been open for three years stating that `PREFER_PARALLEL` doesn't really do what it's supposed to.

What's your experience?

File: opm/simulators/utils/HDF5File.cpp [hunk 2]
```diff
@@ -76,63 +96,71 @@ HDF5File::~HDF5File()
...
-
-    hsize_t size = buffer.size();
-    hsize_t start = 0;
-
-    hid_t space = H5Screate_simple(1, &size, nullptr);
-    hid_t dataset_id = H5Dcreate2(grp, dset.c_str(), H5T_NATIVE_CHAR, space,
-                       H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
-    if (dataset_id == H5I_INVALID_HID) {
-        throw std::runtime_error("HDF5File: Trying to write already existing dataset '" + group + '/' + dset + "'");
+        throw std::runtime_error("Failed to create group '" + realGroup + "'");
```
Comment (bska) [inline opm/simulators/utils/HDF5File.cpp:139]: What happens if a reservoir engineer sees the diagnostic "Failed to create group ..."?  Are they going to associate that with HDF5 or will there be confusion?

**Review by bska — COMMENTED** [review]
Just some high level comments, since I'm not familiar with the HDF5 API.

File: CMakeLists.txt (hunk 1 above)
Comment (akva2) [reply CMakeLists.txt:96]: i haven't had any issues with it on debian derivates. that anaconda creates issues is business as usual.

File: opm/simulators/utils/HDF5File.cpp (hunk 2 above)
Comment (akva2) [reply opm/simulators/utils/HDF5File.cpp:139]: i was a bit torn here. the "issue" is that this now (mostly) used in outer parallel exception handling macros which adds a prefix to the message. So in practice the exception would read "Error saving serialized state: Failed to create group '..'"

File: CMakeLists.txt (hunk 1 above)
Comment (bska) [reply CMakeLists.txt:96]: > i haven't had any issues with it on debian derivates.

Understood.  Then I'm happy to have this.

File: opm/simulators/utils/HDF5File.cpp (hunk 2 above)
Comment (bska) [reply opm/simulators/utils/HDF5File.cpp:139]: > in practice the exception would read "Error saving serialized state: Failed to create group '..'"

Ah, okay.  Then this is fine with me.  We'll merge this now and then adjust diagnostics and the overall implementation as need arises.

**Pull request merged**